spinnaker/keel
# feat(ec2): EBS volumes: default to SSD

# Problem

Occasionally, we see a keel instances get "stuck": they stop actuating for several minutes, and stop reporting metrics. This is always accompanied by a spike in disk load.

One theory is that the problem is related to the older [magnetic][1]  EBS volume types (denoted `"standard"`) that clouddriver defaults to, as defined in [AwsConfiguration.groovy][2].

We'd like to run keel using an SSD-based EBS volume type, such as "gp2", which is now the default AWS volume type.

Unfortunately, we can't just specify volume type in keel's delivery configuration, because the current implementation of LaunchConfigurationSpec does not allow us to specify block device mappings.

# Proposed solution

Change the EBS volume default to "gp2", which is SSD-based, and is the current AWS default volume type.

Note: this change will configure all EC2 clusters to use "gp2" (SSD) for their EBS volume types instead of "standard" (magnetic). 

# Overview of changes

The only behavioral change is in ClusterHandler.kt, where we add a new `blockDevices` field to the payload for the orca job that creates a new server group.

The configuration info for EBS volumes is encapsulated in a new class: BlockDeviceConfig

# Test plan

1. Deploy keel to staging environment
1. Query the volume configuration of an instance from a managed EC2 test app
1. Trigger a deploy of the test app so keel creates a new ASG for the test app
1. Verify that:
   - The mounted EBS volume is of type `gp2`
   - The volume configuration is otherwise unchanged

We can check info of mounted volumes for an instance-id (say, `i-0123456789ABCDEF0`) using the following command:

```bash
$ aws ec2 describe-volumes --region us-east-1 \
 --filters Name=attachment.instance-id,Values=i-0123456789ABCDEF0
```

# FAQ

## Why does this diff include all of that information about volume sizes? The text above doesn't mention volume size at all.

If we want to change the EBS volume type, we need to pass [block device][3] objects. Each object has a "size" field that is required. I tested that the field is required by creating a pipeline where I edited the deploy stage to customize the volume type without specifying the size, like this:

```json
     "blockDevices": [
        {
          "deviceName": "/dev/sdb",
          "volumeType": "gp2"
        }
      ],
```

Alas, the pipeline failed with the following error:

```
EBS device /dev/sdb missing required value size
```

Clouddriver determines the default EBS volume size by instance type. I configured keel to use the same EBS volume sizes as clouddriver uses by copying over the table of values from clouddriver's [BlockDeviceConfig][4] class.

## Why not modify keel to allow users to specify a custom block device mapping instead of changing the default?

It felt more managed-delivery-ish to change an infrastructure detail behind the scenes than to introduce another low-level implementation detail. Everyone running on Managed Delivery now gets SSD-backed volumes! 

## Will this change trigger diffs/actuations for all EC2 clusters when it rolls out?

No, this change shouldn't result in any new "diffs", because we aren't encoding the volume type in the representation of a server group that gets diffed. We're just adding some new parameters to the payload of the CreateServerGroup orca job.

## Why is the volume type called "gp2"?

The "gp" stands for "general purpose", I believe. No idea where the "2" comes from.

# Where I'm looking for feedback

Is this a reasonable approach? Other approaches we could take are:

* Expose block device config in the delivery config instead
* Modify clouddriver so that we can specify a volume type without a size


# References

* [Amazon EBS volume types docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volume-types.html)


[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volume-types.html#EBSVolumeTypes_standard
[2]: https://github.com/spinnaker/clouddriver/blob/a54baac178fe7427ea8589334c84bfa136e0dac4/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/config/AwsConfiguration.groovy#L162
[3]: https://github.com/spinnaker/clouddriver/blob/29c5c1435ab0c750c4b491208f600647a3e98901/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonBlockDevice.groovy
[4]: https://github.com/spinnaker/clouddriver/blob/6338a60c2f6736570d71e8ae7072d3d266f52fff/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/BlockDeviceConfig.groovy
